sound/soc/sof/topology.c

```diff
@@ -1957,21 +1963,37 @@ static int sof_widget_unload(struct snd_soc_component *scomp,
 		/* update enabled cores mask */
 		sdev->enabled_cores_mask &= ~(1 << pipeline->core);
 
-		break;
-	case snd_soc_dapm_pga:
-
-		/* get volume kcontrol */
-		kc = &widget->kcontrol_news[0];
-		sm = (struct soc_mixer_control *)kc->private_value;
-		scontrol = sm->dobj.private;
-
-		/* free volume table */
-		kfree(scontrol->volume_table);
 		break;
 	default:
 		break;
 	}
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+		switch (dobj->widget.kcontrol_type) {
+		case SND_SOC_TPLG_TYPE_MIXER:
+			sm = (struct soc_mixer_control *)kc->private_value;
+			scontrol = sm->dobj.private;
+			if (sm->max > 1)
```

Comment:
@bardliao Thanks. This PR looks good for what it was intended. But, in reality we never actually make sure that the volume table is set up only if sm->max is > 1. I think we should fix that.

Reply:
@ranj063 I am thinking should we move set_up_volume_table() from sof_widget_load_pga() to sof_control_load_volume()? It seems to make more sense to me.

Reply:
@bardliao yes, makes sense. I have no idea why I put it inside load_pga() :)

Reply:
@ranj063 I may know the reason why you put it inside load_pga(). soc_tplg_create_tlv() is called after soc_tplg_init_kcontrol(), so you can't get tlv.p in sof_control_load().  I think there should be no side effect if we move soc_tplg_create_tlv() to the front of soc_tplg_init_kcontrol(). Then we can move set_up_volume_table() to sof_control_load_volume().

Reply:
@bardliao yes, you are right. moving soc_tplg_create_tlv() ahead of soc_tplg_init_kcontrol() would fix the problem. If you look into soc-topology.c, the call to soc_tplg_create_tlv() inside soc_tplg_dapm_widget_dmixer_create() is what is of significance to us. I added the called to soc_tplg_create_tlv() after soc_tplg_init_kcontrol() to be consistent with soc_tplg_dmixer_create(). So lets change it both places.